# Libraries

In [4]:
# uncomment this if you dont have tabnet installed
# !pip install pytorch-tabnet

In [62]:
# import libraries
import pandas as pd
import numpy as np
import torch
import matplotlib.pyplot as plt

from pytorch_tabnet.tab_model import TabNetClassifier
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import roc_auc_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_recall_curve
from sklearn.metrics import average_precision_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split

In [6]:
# for google colab
from google.colab import drive
drive.mount('/content/drive')

device = "cuda" if torch.cuda.is_available() else "cpu"
print("Using device:", device)

Mounted at /content/drive
Using device: cuda


# Read in Files

In [8]:
# load csv files
train_dataset = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/datasets/train.csv')
test_dataset = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/datasets/test.csv')

# train (cross-validation)
X = train_dataset.drop(columns=["hit"]).to_numpy()
y = train_dataset['hit'].to_numpy()

# test (not used in cross-validation)
X_test = test_dataset.drop(columns=["hit"]).to_numpy()
y_test = test_dataset["hit"].to_numpy()


In [9]:
print("===== TRAINING DATASET =====")
print(f"Full train shape: {train_dataset.shape}")
print(f"X_train shape: {X.shape}")
print(f"y_train shape: {y.shape}")

print("\n===== TEST DATASET =====")
print(f"Full test shape: {test_dataset.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_test shape: {y_test.shape}")

===== TRAINING DATASET =====
Full train shape: (19348, 31)
X_train shape: (19348, 30)
y_train shape: (19348,)

===== TEST DATASET =====
Full test shape: (4837, 31)
X_test shape: (4837, 30)
y_test shape: (4837,)


# TabNet with Cross Validation

## Train with Cross-Validation

In [38]:
configs = [
    dict(n_d=32, n_a=32, n_steps=5, gamma=1.5, lambda_sparse=0.001, lr=0.001),
    dict(n_d=32, n_a=32, n_steps=6, gamma=1.5, lambda_sparse=0.001, lr=0.001),
    dict(n_d=32, n_a=32, n_steps=5, gamma=1.8, lambda_sparse=0.01, lr=0.001),
    dict(n_d=32, n_a=32, n_steps=5, gamma=1.5, lambda_sparse=0,      lr=0.001),
    dict(n_d=64, n_a=64, n_steps=5, gamma=1.5, lambda_sparse=0.001, lr=0.001),
]

# n_d: numner of units in the decision prediction layer

# n_a: number of units in the attention layer
# (how the model selects important features at each step)

# n_steps: number of sequential steps,
# (how many times the model applies attention + decision blocks)

# gamma: controls how many features can be reused across steps

# lambda_sparse: Encourages the model to use fewer features at each step.

# lr: step size used by the optimizer when updating weights


skf = StratifiedKFold(n_splits=5, random_state=426, shuffle=True)
results = []

for ci, cfg in enumerate(configs, 1):
    fold_aucs = []

    for fold, (train_idx, val_idx) in enumerate(skf.split(X, y), 1):
        # split
        X_train, X_val = X[train_idx], X[val_idx]
        y_train, y_val = y[train_idx], y[val_idx]

        # model from config
        tb_cls = TabNetClassifier(
            n_d=cfg["n_d"],
            n_a=cfg["n_a"],
            n_steps=cfg["n_steps"],
            gamma=cfg["gamma"],
            lambda_sparse=cfg["lambda_sparse"],
            optimizer_fn=torch.optim.AdamW,
            optimizer_params=dict(lr=cfg["lr"], weight_decay=1e-5),
            mask_type="entmax",
            device_name=device
        )

        # train
        tb_cls.fit(
            X_train, y_train,
            eval_set=[(X_val, y_val)],
            eval_name=['val'],
            eval_metric=["auc"],
            max_epochs=100,
            patience=10,
            batch_size=1024,
            virtual_batch_size=128,
            drop_last=False
        )

        # AUC from probabilities
        val_proba = tb_cls.predict_proba(X_val)[:, 1]
        auc = roc_auc_score(y_val, val_proba)
        fold_aucs.append(auc)

        print(f"Config {ci}/{len(configs)} | Fold {fold} AUC: {auc:.4f}")

    mean_auc = float(np.mean(fold_aucs))
    std_auc  = float(np.std(fold_aucs))

    results.append((ci, cfg, mean_auc, std_auc))

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.67017 | val_auc: 0.48718 |  0:00:00s
epoch 1  | loss: 0.56115 | val_auc: 0.49323 |  0:00:00s
epoch 2  | loss: 0.52265 | val_auc: 0.48477 |  0:00:01s
epoch 3  | loss: 0.50355 | val_auc: 0.5032  |  0:00:01s
epoch 4  | loss: 0.48772 | val_auc: 0.50643 |  0:00:01s
epoch 5  | loss: 0.47809 | val_auc: 0.52686 |  0:00:02s
epoch 6  | loss: 0.46246 | val_auc: 0.53695 |  0:00:02s
epoch 7  | loss: 0.45326 | val_auc: 0.56738 |  0:00:02s
epoch 8  | loss: 0.44707 | val_auc: 0.58186 |  0:00:03s
epoch 9  | loss: 0.43646 | val_auc: 0.59923 |  0:00:03s
epoch 10 | loss: 0.43471 | val_auc: 0.60065 |  0:00:04s
epoch 11 | loss: 0.42872 | val_auc: 0.61584 |  0:00:04s
epoch 12 | loss: 0.42493 | val_auc: 0.6224  |  0:00:04s
epoch 13 | loss: 0.42203 | val_auc: 0.63172 |  0:00:05s
epoch 14 | loss: 0.4206  | val_auc: 0.63652 |  0:00:05s
epoch 15 | loss: 0.41746 | val_auc: 0.64944 |  0:00:05s
epoch 16 | loss: 0.42012 | val_auc: 0.65831 |  0:00:06s
epoch 17 | loss: 0.41308 | val_auc: 0.65066 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.66281 | val_auc: 0.49306 |  0:00:00s
epoch 1  | loss: 0.57578 | val_auc: 0.48933 |  0:00:00s
epoch 2  | loss: 0.53071 | val_auc: 0.50055 |  0:00:01s
epoch 3  | loss: 0.49772 | val_auc: 0.50534 |  0:00:01s
epoch 4  | loss: 0.49063 | val_auc: 0.50975 |  0:00:01s
epoch 5  | loss: 0.46425 | val_auc: 0.53435 |  0:00:02s
epoch 6  | loss: 0.46147 | val_auc: 0.54597 |  0:00:02s
epoch 7  | loss: 0.45249 | val_auc: 0.56084 |  0:00:02s
epoch 8  | loss: 0.44361 | val_auc: 0.57048 |  0:00:03s
epoch 9  | loss: 0.44005 | val_auc: 0.58253 |  0:00:03s
epoch 10 | loss: 0.43146 | val_auc: 0.59419 |  0:00:03s
epoch 11 | loss: 0.4275  | val_auc: 0.59456 |  0:00:04s
epoch 12 | loss: 0.43052 | val_auc: 0.60413 |  0:00:04s
epoch 13 | loss: 0.42349 | val_auc: 0.62065 |  0:00:05s
epoch 14 | loss: 0.41932 | val_auc: 0.63407 |  0:00:05s
epoch 15 | loss: 0.41824 | val_auc: 0.63233 |  0:00:05s
epoch 16 | loss: 0.41224 | val_auc: 0.62534 |  0:00:06s
epoch 17 | loss: 0.41221 | val_auc: 0.63262 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.67958 | val_auc: 0.46361 |  0:00:00s
epoch 1  | loss: 0.57089 | val_auc: 0.46889 |  0:00:00s
epoch 2  | loss: 0.52432 | val_auc: 0.46774 |  0:00:01s
epoch 3  | loss: 0.50236 | val_auc: 0.47003 |  0:00:01s
epoch 4  | loss: 0.48217 | val_auc: 0.49442 |  0:00:01s
epoch 5  | loss: 0.46402 | val_auc: 0.52195 |  0:00:02s
epoch 6  | loss: 0.45408 | val_auc: 0.54109 |  0:00:02s
epoch 7  | loss: 0.44481 | val_auc: 0.57425 |  0:00:02s
epoch 8  | loss: 0.43785 | val_auc: 0.60036 |  0:00:03s
epoch 9  | loss: 0.43486 | val_auc: 0.6168  |  0:00:03s
epoch 10 | loss: 0.43021 | val_auc: 0.62475 |  0:00:03s
epoch 11 | loss: 0.42789 | val_auc: 0.63095 |  0:00:04s
epoch 12 | loss: 0.42555 | val_auc: 0.64601 |  0:00:04s
epoch 13 | loss: 0.42277 | val_auc: 0.65328 |  0:00:05s
epoch 14 | loss: 0.42021 | val_auc: 0.6649  |  0:00:05s
epoch 15 | loss: 0.41974 | val_auc: 0.663   |  0:00:05s
epoch 16 | loss: 0.41427 | val_auc: 0.65869 |  0:00:06s
epoch 17 | loss: 0.41333 | val_auc: 0.66828 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


Config 1/5 | Fold 3 AUC: 0.6703
epoch 0  | loss: 0.67227 | val_auc: 0.47986 |  0:00:00s
epoch 1  | loss: 0.57669 | val_auc: 0.48358 |  0:00:00s
epoch 2  | loss: 0.52941 | val_auc: 0.48338 |  0:00:01s
epoch 3  | loss: 0.49685 | val_auc: 0.51663 |  0:00:01s
epoch 4  | loss: 0.48232 | val_auc: 0.53519 |  0:00:01s
epoch 5  | loss: 0.46018 | val_auc: 0.54529 |  0:00:02s
epoch 6  | loss: 0.45573 | val_auc: 0.55223 |  0:00:02s
epoch 7  | loss: 0.44812 | val_auc: 0.5667  |  0:00:02s
epoch 8  | loss: 0.4408  | val_auc: 0.5814  |  0:00:03s
epoch 9  | loss: 0.4355  | val_auc: 0.59639 |  0:00:03s
epoch 10 | loss: 0.42726 | val_auc: 0.61935 |  0:00:03s
epoch 11 | loss: 0.42583 | val_auc: 0.61917 |  0:00:04s
epoch 12 | loss: 0.42274 | val_auc: 0.61929 |  0:00:04s
epoch 13 | loss: 0.42303 | val_auc: 0.62661 |  0:00:05s
epoch 14 | loss: 0.41558 | val_auc: 0.62112 |  0:00:05s
epoch 15 | loss: 0.41768 | val_auc: 0.62021 |  0:00:05s
epoch 16 | loss: 0.41236 | val_auc: 0.61773 |  0:00:06s
epoch 17 | loss:

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


Config 1/5 | Fold 4 AUC: 0.6700
epoch 0  | loss: 0.66988 | val_auc: 0.49034 |  0:00:00s
epoch 1  | loss: 0.57199 | val_auc: 0.48833 |  0:00:00s
epoch 2  | loss: 0.53488 | val_auc: 0.49931 |  0:00:01s
epoch 3  | loss: 0.5072  | val_auc: 0.51081 |  0:00:01s
epoch 4  | loss: 0.48264 | val_auc: 0.5363  |  0:00:01s
epoch 5  | loss: 0.47079 | val_auc: 0.563   |  0:00:02s
epoch 6  | loss: 0.46313 | val_auc: 0.56601 |  0:00:02s
epoch 7  | loss: 0.44884 | val_auc: 0.559   |  0:00:02s
epoch 8  | loss: 0.44202 | val_auc: 0.5799  |  0:00:03s
epoch 9  | loss: 0.43742 | val_auc: 0.59115 |  0:00:03s
epoch 10 | loss: 0.43302 | val_auc: 0.60458 |  0:00:04s
epoch 11 | loss: 0.43178 | val_auc: 0.61246 |  0:00:04s
epoch 12 | loss: 0.42549 | val_auc: 0.61372 |  0:00:04s
epoch 13 | loss: 0.42146 | val_auc: 0.61209 |  0:00:05s
epoch 14 | loss: 0.41851 | val_auc: 0.61968 |  0:00:05s
epoch 15 | loss: 0.41948 | val_auc: 0.60383 |  0:00:05s
epoch 16 | loss: 0.41714 | val_auc: 0.62119 |  0:00:06s
epoch 17 | loss:

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 1.17388 | val_auc: 0.53343 |  0:00:00s
epoch 1  | loss: 0.65758 | val_auc: 0.49826 |  0:00:00s
epoch 2  | loss: 0.57283 | val_auc: 0.50795 |  0:00:01s
epoch 3  | loss: 0.55388 | val_auc: 0.51074 |  0:00:01s
epoch 4  | loss: 0.52418 | val_auc: 0.53817 |  0:00:01s
epoch 5  | loss: 0.50855 | val_auc: 0.52553 |  0:00:02s
epoch 6  | loss: 0.48985 | val_auc: 0.54753 |  0:00:02s
epoch 7  | loss: 0.47304 | val_auc: 0.54847 |  0:00:03s
epoch 8  | loss: 0.47184 | val_auc: 0.5503  |  0:00:03s
epoch 9  | loss: 0.45583 | val_auc: 0.56687 |  0:00:03s
epoch 10 | loss: 0.44664 | val_auc: 0.5813  |  0:00:04s
epoch 11 | loss: 0.43906 | val_auc: 0.58322 |  0:00:04s
epoch 12 | loss: 0.4388  | val_auc: 0.59131 |  0:00:05s
epoch 13 | loss: 0.43581 | val_auc: 0.59442 |  0:00:05s
epoch 14 | loss: 0.43482 | val_auc: 0.59194 |  0:00:05s
epoch 15 | loss: 0.42461 | val_auc: 0.59479 |  0:00:06s
epoch 16 | loss: 0.42798 | val_auc: 0.59467 |  0:00:06s
epoch 17 | loss: 0.42481 | val_auc: 0.59666 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 2/5 | Fold 1 AUC: 0.6797


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 1.18885 | val_auc: 0.52144 |  0:00:00s
epoch 1  | loss: 0.66907 | val_auc: 0.50917 |  0:00:00s
epoch 2  | loss: 0.5793  | val_auc: 0.50479 |  0:00:01s
epoch 3  | loss: 0.55477 | val_auc: 0.52792 |  0:00:01s
epoch 4  | loss: 0.52763 | val_auc: 0.51881 |  0:00:01s
epoch 5  | loss: 0.5046  | val_auc: 0.52119 |  0:00:02s
epoch 6  | loss: 0.48936 | val_auc: 0.52939 |  0:00:02s
epoch 7  | loss: 0.48385 | val_auc: 0.55216 |  0:00:03s
epoch 8  | loss: 0.47599 | val_auc: 0.55382 |  0:00:03s
epoch 9  | loss: 0.46264 | val_auc: 0.56761 |  0:00:03s
epoch 10 | loss: 0.45322 | val_auc: 0.57744 |  0:00:04s
epoch 11 | loss: 0.44867 | val_auc: 0.59531 |  0:00:04s
epoch 12 | loss: 0.44027 | val_auc: 0.59681 |  0:00:05s
epoch 13 | loss: 0.43941 | val_auc: 0.59919 |  0:00:05s
epoch 14 | loss: 0.43558 | val_auc: 0.61111 |  0:00:05s
epoch 15 | loss: 0.43343 | val_auc: 0.62223 |  0:00:06s
epoch 16 | loss: 0.43063 | val_auc: 0.60879 |  0:00:06s
epoch 17 | loss: 0.42788 | val_auc: 0.60684 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 2/5 | Fold 2 AUC: 0.6792


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 1.21514 | val_auc: 0.51271 |  0:00:00s
epoch 1  | loss: 0.67531 | val_auc: 0.53211 |  0:00:00s
epoch 2  | loss: 0.56815 | val_auc: 0.5161  |  0:00:01s
epoch 3  | loss: 0.54532 | val_auc: 0.51778 |  0:00:01s
epoch 4  | loss: 0.51096 | val_auc: 0.51045 |  0:00:02s
epoch 5  | loss: 0.4969  | val_auc: 0.49551 |  0:00:02s
epoch 6  | loss: 0.48426 | val_auc: 0.49437 |  0:00:02s
epoch 7  | loss: 0.47364 | val_auc: 0.51144 |  0:00:03s
epoch 8  | loss: 0.46067 | val_auc: 0.52596 |  0:00:03s
epoch 9  | loss: 0.45593 | val_auc: 0.53692 |  0:00:04s
epoch 10 | loss: 0.45212 | val_auc: 0.5524  |  0:00:04s
epoch 11 | loss: 0.4484  | val_auc: 0.56623 |  0:00:04s
epoch 12 | loss: 0.44738 | val_auc: 0.56912 |  0:00:05s
epoch 13 | loss: 0.4369  | val_auc: 0.5708  |  0:00:05s
epoch 14 | loss: 0.43773 | val_auc: 0.58111 |  0:00:06s
epoch 15 | loss: 0.42745 | val_auc: 0.58918 |  0:00:06s
epoch 16 | loss: 0.43232 | val_auc: 0.59995 |  0:00:06s
epoch 17 | loss: 0.42667 | val_auc: 0.60802 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 2/5 | Fold 3 AUC: 0.6447


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 1.20565 | val_auc: 0.5132  |  0:00:00s
epoch 1  | loss: 0.68522 | val_auc: 0.52938 |  0:00:00s
epoch 2  | loss: 0.57128 | val_auc: 0.51869 |  0:00:01s
epoch 3  | loss: 0.53702 | val_auc: 0.52143 |  0:00:01s
epoch 4  | loss: 0.51503 | val_auc: 0.5053  |  0:00:02s
epoch 5  | loss: 0.49547 | val_auc: 0.52707 |  0:00:02s
epoch 6  | loss: 0.47769 | val_auc: 0.5312  |  0:00:03s
epoch 7  | loss: 0.46638 | val_auc: 0.54728 |  0:00:03s
epoch 8  | loss: 0.46282 | val_auc: 0.54428 |  0:00:04s
epoch 9  | loss: 0.4495  | val_auc: 0.55992 |  0:00:04s
epoch 10 | loss: 0.44244 | val_auc: 0.58196 |  0:00:04s
epoch 11 | loss: 0.44044 | val_auc: 0.58073 |  0:00:05s
epoch 12 | loss: 0.44248 | val_auc: 0.57816 |  0:00:05s
epoch 13 | loss: 0.43426 | val_auc: 0.58279 |  0:00:06s
epoch 14 | loss: 0.42856 | val_auc: 0.58401 |  0:00:06s
epoch 15 | loss: 0.42729 | val_auc: 0.59222 |  0:00:07s
epoch 16 | loss: 0.42904 | val_auc: 0.58203 |  0:00:07s
epoch 17 | loss: 0.42717 | val_auc: 0.58736 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 2/5 | Fold 4 AUC: 0.6595


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 1.18529 | val_auc: 0.5323  |  0:00:00s
epoch 1  | loss: 0.68681 | val_auc: 0.50746 |  0:00:00s
epoch 2  | loss: 0.59024 | val_auc: 0.50279 |  0:00:01s
epoch 3  | loss: 0.55189 | val_auc: 0.52455 |  0:00:01s
epoch 4  | loss: 0.52618 | val_auc: 0.52387 |  0:00:01s
epoch 5  | loss: 0.49587 | val_auc: 0.52902 |  0:00:02s
epoch 6  | loss: 0.4941  | val_auc: 0.53562 |  0:00:02s
epoch 7  | loss: 0.46782 | val_auc: 0.52639 |  0:00:03s
epoch 8  | loss: 0.46303 | val_auc: 0.54057 |  0:00:03s
epoch 9  | loss: 0.46181 | val_auc: 0.5637  |  0:00:03s
epoch 10 | loss: 0.45168 | val_auc: 0.56309 |  0:00:04s
epoch 11 | loss: 0.45052 | val_auc: 0.56917 |  0:00:04s
epoch 12 | loss: 0.43884 | val_auc: 0.58625 |  0:00:05s
epoch 13 | loss: 0.44084 | val_auc: 0.59053 |  0:00:05s
epoch 14 | loss: 0.43932 | val_auc: 0.59703 |  0:00:05s
epoch 15 | loss: 0.43099 | val_auc: 0.60432 |  0:00:06s
epoch 16 | loss: 0.43253 | val_auc: 0.61067 |  0:00:06s
epoch 17 | loss: 0.42475 | val_auc: 0.61625 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 2/5 | Fold 5 AUC: 0.6635


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.6881  | val_auc: 0.49882 |  0:00:00s
epoch 1  | loss: 0.58038 | val_auc: 0.49073 |  0:00:00s
epoch 2  | loss: 0.54889 | val_auc: 0.50838 |  0:00:01s
epoch 3  | loss: 0.51949 | val_auc: 0.47328 |  0:00:01s
epoch 4  | loss: 0.49394 | val_auc: 0.51466 |  0:00:01s
epoch 5  | loss: 0.48339 | val_auc: 0.52569 |  0:00:02s
epoch 6  | loss: 0.46811 | val_auc: 0.52226 |  0:00:02s
epoch 7  | loss: 0.46389 | val_auc: 0.53902 |  0:00:02s
epoch 8  | loss: 0.45761 | val_auc: 0.54321 |  0:00:03s
epoch 9  | loss: 0.44798 | val_auc: 0.5569  |  0:00:03s
epoch 10 | loss: 0.44394 | val_auc: 0.58119 |  0:00:04s
epoch 11 | loss: 0.44337 | val_auc: 0.5971  |  0:00:04s
epoch 12 | loss: 0.43079 | val_auc: 0.60443 |  0:00:04s
epoch 13 | loss: 0.4383  | val_auc: 0.60783 |  0:00:05s
epoch 14 | loss: 0.42882 | val_auc: 0.61102 |  0:00:05s
epoch 15 | loss: 0.42785 | val_auc: 0.61344 |  0:00:05s
epoch 16 | loss: 0.42401 | val_auc: 0.62646 |  0:00:06s
epoch 17 | loss: 0.42383 | val_auc: 0.63184 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.67661 | val_auc: 0.47257 |  0:00:00s
epoch 1  | loss: 0.5893  | val_auc: 0.48131 |  0:00:00s
epoch 2  | loss: 0.54454 | val_auc: 0.49142 |  0:00:01s
epoch 3  | loss: 0.52559 | val_auc: 0.49492 |  0:00:01s
epoch 4  | loss: 0.49538 | val_auc: 0.4996  |  0:00:01s
epoch 5  | loss: 0.48335 | val_auc: 0.5164  |  0:00:02s
epoch 6  | loss: 0.47296 | val_auc: 0.5373  |  0:00:02s
epoch 7  | loss: 0.46716 | val_auc: 0.54887 |  0:00:02s
epoch 8  | loss: 0.45744 | val_auc: 0.5733  |  0:00:03s
epoch 9  | loss: 0.45002 | val_auc: 0.58138 |  0:00:03s
epoch 10 | loss: 0.44881 | val_auc: 0.57391 |  0:00:03s
epoch 11 | loss: 0.44534 | val_auc: 0.58088 |  0:00:04s
epoch 12 | loss: 0.44118 | val_auc: 0.57614 |  0:00:04s
epoch 13 | loss: 0.44026 | val_auc: 0.58841 |  0:00:04s
epoch 14 | loss: 0.43562 | val_auc: 0.59317 |  0:00:05s
epoch 15 | loss: 0.43322 | val_auc: 0.59542 |  0:00:05s
epoch 16 | loss: 0.43141 | val_auc: 0.61812 |  0:00:05s
epoch 17 | loss: 0.42883 | val_auc: 0.60175 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


Config 3/5 | Fold 2 AUC: 0.6544
epoch 0  | loss: 0.684   | val_auc: 0.48697 |  0:00:00s
epoch 1  | loss: 0.591   | val_auc: 0.4846  |  0:00:00s
epoch 2  | loss: 0.53451 | val_auc: 0.46977 |  0:00:01s
epoch 3  | loss: 0.51599 | val_auc: 0.4751  |  0:00:01s
epoch 4  | loss: 0.49946 | val_auc: 0.4795  |  0:00:01s
epoch 5  | loss: 0.4876  | val_auc: 0.51162 |  0:00:02s
epoch 6  | loss: 0.48208 | val_auc: 0.53392 |  0:00:02s
epoch 7  | loss: 0.45715 | val_auc: 0.55611 |  0:00:02s
epoch 8  | loss: 0.4556  | val_auc: 0.59403 |  0:00:03s
epoch 9  | loss: 0.45023 | val_auc: 0.61443 |  0:00:03s
epoch 10 | loss: 0.44552 | val_auc: 0.6165  |  0:00:04s
epoch 11 | loss: 0.44644 | val_auc: 0.6196  |  0:00:04s
epoch 12 | loss: 0.43243 | val_auc: 0.61532 |  0:00:04s
epoch 13 | loss: 0.43454 | val_auc: 0.62919 |  0:00:05s
epoch 14 | loss: 0.42662 | val_auc: 0.62441 |  0:00:05s
epoch 15 | loss: 0.42947 | val_auc: 0.63302 |  0:00:05s
epoch 16 | loss: 0.42889 | val_auc: 0.63268 |  0:00:06s
epoch 17 | loss:

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.6956  | val_auc: 0.49071 |  0:00:00s
epoch 1  | loss: 0.5899  | val_auc: 0.51013 |  0:00:00s
epoch 2  | loss: 0.54638 | val_auc: 0.50608 |  0:00:01s
epoch 3  | loss: 0.51189 | val_auc: 0.51585 |  0:00:01s
epoch 4  | loss: 0.49673 | val_auc: 0.51383 |  0:00:01s
epoch 5  | loss: 0.48012 | val_auc: 0.53023 |  0:00:02s
epoch 6  | loss: 0.47027 | val_auc: 0.53922 |  0:00:02s
epoch 7  | loss: 0.4688  | val_auc: 0.54161 |  0:00:03s
epoch 8  | loss: 0.45691 | val_auc: 0.54352 |  0:00:03s
epoch 9  | loss: 0.44241 | val_auc: 0.57238 |  0:00:04s
epoch 10 | loss: 0.44778 | val_auc: 0.57735 |  0:00:04s
epoch 11 | loss: 0.44371 | val_auc: 0.59247 |  0:00:04s
epoch 12 | loss: 0.43735 | val_auc: 0.60313 |  0:00:05s
epoch 13 | loss: 0.43396 | val_auc: 0.60644 |  0:00:05s
epoch 14 | loss: 0.4301  | val_auc: 0.61165 |  0:00:05s
epoch 15 | loss: 0.42986 | val_auc: 0.61567 |  0:00:06s
epoch 16 | loss: 0.42481 | val_auc: 0.61667 |  0:00:06s
epoch 17 | loss: 0.42721 | val_auc: 0.61718 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.67163 | val_auc: 0.48447 |  0:00:00s
epoch 1  | loss: 0.5955  | val_auc: 0.51913 |  0:00:00s
epoch 2  | loss: 0.55821 | val_auc: 0.50691 |  0:00:01s
epoch 3  | loss: 0.52582 | val_auc: 0.52773 |  0:00:01s
epoch 4  | loss: 0.48978 | val_auc: 0.54896 |  0:00:01s
epoch 5  | loss: 0.4844  | val_auc: 0.55554 |  0:00:02s
epoch 6  | loss: 0.47863 | val_auc: 0.55804 |  0:00:02s
epoch 7  | loss: 0.46174 | val_auc: 0.5686  |  0:00:02s
epoch 8  | loss: 0.4509  | val_auc: 0.56751 |  0:00:03s
epoch 9  | loss: 0.44795 | val_auc: 0.56925 |  0:00:03s
epoch 10 | loss: 0.44505 | val_auc: 0.57343 |  0:00:03s
epoch 11 | loss: 0.44588 | val_auc: 0.57818 |  0:00:04s
epoch 12 | loss: 0.44359 | val_auc: 0.57907 |  0:00:04s
epoch 13 | loss: 0.43478 | val_auc: 0.5965  |  0:00:04s
epoch 14 | loss: 0.43344 | val_auc: 0.60848 |  0:00:05s
epoch 15 | loss: 0.42866 | val_auc: 0.62028 |  0:00:05s
epoch 16 | loss: 0.42534 | val_auc: 0.63036 |  0:00:06s
epoch 17 | loss: 0.42535 | val_auc: 0.63401 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 3/5 | Fold 5 AUC: 0.6831


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.66443 | val_auc: 0.4882  |  0:00:00s
epoch 1  | loss: 0.56845 | val_auc: 0.49592 |  0:00:00s
epoch 2  | loss: 0.53849 | val_auc: 0.51931 |  0:00:01s
epoch 3  | loss: 0.50796 | val_auc: 0.5127  |  0:00:01s
epoch 4  | loss: 0.48403 | val_auc: 0.52687 |  0:00:01s
epoch 5  | loss: 0.46928 | val_auc: 0.52943 |  0:00:02s
epoch 6  | loss: 0.46107 | val_auc: 0.53174 |  0:00:02s
epoch 7  | loss: 0.45072 | val_auc: 0.55725 |  0:00:02s
epoch 8  | loss: 0.44899 | val_auc: 0.57758 |  0:00:03s
epoch 9  | loss: 0.43486 | val_auc: 0.59116 |  0:00:03s
epoch 10 | loss: 0.4302  | val_auc: 0.59294 |  0:00:04s
epoch 11 | loss: 0.42614 | val_auc: 0.60164 |  0:00:04s
epoch 12 | loss: 0.42327 | val_auc: 0.61284 |  0:00:04s
epoch 13 | loss: 0.41732 | val_auc: 0.63063 |  0:00:05s
epoch 14 | loss: 0.41666 | val_auc: 0.62383 |  0:00:05s
epoch 15 | loss: 0.41517 | val_auc: 0.63382 |  0:00:05s
epoch 16 | loss: 0.41548 | val_auc: 0.63358 |  0:00:06s
epoch 17 | loss: 0.41434 | val_auc: 0.64424 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 4/5 | Fold 1 AUC: 0.6645


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.65495 | val_auc: 0.48068 |  0:00:00s
epoch 1  | loss: 0.57107 | val_auc: 0.47759 |  0:00:00s
epoch 2  | loss: 0.5303  | val_auc: 0.48636 |  0:00:01s
epoch 3  | loss: 0.49822 | val_auc: 0.49944 |  0:00:01s
epoch 4  | loss: 0.47774 | val_auc: 0.50238 |  0:00:01s
epoch 5  | loss: 0.46593 | val_auc: 0.52379 |  0:00:02s
epoch 6  | loss: 0.45491 | val_auc: 0.54655 |  0:00:02s
epoch 7  | loss: 0.45046 | val_auc: 0.55349 |  0:00:02s
epoch 8  | loss: 0.44464 | val_auc: 0.58001 |  0:00:03s
epoch 9  | loss: 0.43279 | val_auc: 0.60063 |  0:00:03s
epoch 10 | loss: 0.42631 | val_auc: 0.60623 |  0:00:04s
epoch 11 | loss: 0.42366 | val_auc: 0.61746 |  0:00:04s
epoch 12 | loss: 0.41835 | val_auc: 0.61985 |  0:00:04s
epoch 13 | loss: 0.41936 | val_auc: 0.6189  |  0:00:05s
epoch 14 | loss: 0.41641 | val_auc: 0.61825 |  0:00:05s
epoch 15 | loss: 0.41314 | val_auc: 0.61413 |  0:00:05s
epoch 16 | loss: 0.41264 | val_auc: 0.61254 |  0:00:06s
epoch 17 | loss: 0.41086 | val_auc: 0.61433 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


Config 4/5 | Fold 2 AUC: 0.6685
epoch 0  | loss: 0.66608 | val_auc: 0.47149 |  0:00:00s
epoch 1  | loss: 0.57    | val_auc: 0.46249 |  0:00:00s
epoch 2  | loss: 0.52921 | val_auc: 0.46124 |  0:00:01s
epoch 3  | loss: 0.50779 | val_auc: 0.49623 |  0:00:01s
epoch 4  | loss: 0.4826  | val_auc: 0.51826 |  0:00:01s
epoch 5  | loss: 0.46528 | val_auc: 0.55037 |  0:00:02s
epoch 6  | loss: 0.45086 | val_auc: 0.58245 |  0:00:02s
epoch 7  | loss: 0.44467 | val_auc: 0.5896  |  0:00:02s
epoch 8  | loss: 0.44638 | val_auc: 0.59397 |  0:00:03s
epoch 9  | loss: 0.43248 | val_auc: 0.60741 |  0:00:03s
epoch 10 | loss: 0.43073 | val_auc: 0.61653 |  0:00:03s
epoch 11 | loss: 0.43057 | val_auc: 0.61355 |  0:00:04s
epoch 12 | loss: 0.42312 | val_auc: 0.63508 |  0:00:04s
epoch 13 | loss: 0.42252 | val_auc: 0.64687 |  0:00:04s
epoch 14 | loss: 0.41655 | val_auc: 0.65569 |  0:00:05s
epoch 15 | loss: 0.41081 | val_auc: 0.6568  |  0:00:05s
epoch 16 | loss: 0.41448 | val_auc: 0.65929 |  0:00:05s
epoch 17 | loss:

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


Config 4/5 | Fold 3 AUC: 0.6835
epoch 0  | loss: 0.67437 | val_auc: 0.49601 |  0:00:00s
epoch 1  | loss: 0.57369 | val_auc: 0.51422 |  0:00:00s
epoch 2  | loss: 0.52937 | val_auc: 0.52286 |  0:00:01s
epoch 3  | loss: 0.50782 | val_auc: 0.52369 |  0:00:01s
epoch 4  | loss: 0.47942 | val_auc: 0.5203  |  0:00:01s
epoch 5  | loss: 0.47052 | val_auc: 0.51759 |  0:00:02s
epoch 6  | loss: 0.45693 | val_auc: 0.55145 |  0:00:02s
epoch 7  | loss: 0.45667 | val_auc: 0.57616 |  0:00:02s
epoch 8  | loss: 0.43952 | val_auc: 0.59485 |  0:00:03s
epoch 9  | loss: 0.4391  | val_auc: 0.6092  |  0:00:03s
epoch 10 | loss: 0.42822 | val_auc: 0.62415 |  0:00:04s
epoch 11 | loss: 0.43052 | val_auc: 0.62468 |  0:00:04s
epoch 12 | loss: 0.42258 | val_auc: 0.62392 |  0:00:04s
epoch 13 | loss: 0.42056 | val_auc: 0.62312 |  0:00:05s
epoch 14 | loss: 0.42216 | val_auc: 0.62272 |  0:00:05s
epoch 15 | loss: 0.42166 | val_auc: 0.62376 |  0:00:05s
epoch 16 | loss: 0.4094  | val_auc: 0.63299 |  0:00:06s
epoch 17 | loss:

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 4/5 | Fold 4 AUC: 0.6597


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.66378 | val_auc: 0.4935  |  0:00:00s
epoch 1  | loss: 0.57454 | val_auc: 0.49137 |  0:00:00s
epoch 2  | loss: 0.532   | val_auc: 0.5064  |  0:00:01s
epoch 3  | loss: 0.49716 | val_auc: 0.53172 |  0:00:01s
epoch 4  | loss: 0.47622 | val_auc: 0.55725 |  0:00:01s
epoch 5  | loss: 0.46349 | val_auc: 0.55053 |  0:00:02s
epoch 6  | loss: 0.45246 | val_auc: 0.56183 |  0:00:02s
epoch 7  | loss: 0.44595 | val_auc: 0.57417 |  0:00:02s
epoch 8  | loss: 0.43996 | val_auc: 0.58163 |  0:00:03s
epoch 9  | loss: 0.43728 | val_auc: 0.60357 |  0:00:03s
epoch 10 | loss: 0.43097 | val_auc: 0.61851 |  0:00:03s
epoch 11 | loss: 0.42743 | val_auc: 0.62575 |  0:00:04s
epoch 12 | loss: 0.42472 | val_auc: 0.62521 |  0:00:04s
epoch 13 | loss: 0.42165 | val_auc: 0.63063 |  0:00:04s
epoch 14 | loss: 0.41594 | val_auc: 0.62844 |  0:00:05s
epoch 15 | loss: 0.41633 | val_auc: 0.63014 |  0:00:05s
epoch 16 | loss: 0.41208 | val_auc: 0.64436 |  0:00:05s
epoch 17 | loss: 0.40907 | val_auc: 0.64991 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.70339 | val_auc: 0.50149 |  0:00:00s
epoch 1  | loss: 0.58347 | val_auc: 0.47914 |  0:00:00s
epoch 2  | loss: 0.52437 | val_auc: 0.50343 |  0:00:01s
epoch 3  | loss: 0.494   | val_auc: 0.52186 |  0:00:01s
epoch 4  | loss: 0.47509 | val_auc: 0.55077 |  0:00:01s
epoch 5  | loss: 0.45676 | val_auc: 0.56743 |  0:00:02s
epoch 6  | loss: 0.45149 | val_auc: 0.58987 |  0:00:02s
epoch 7  | loss: 0.44284 | val_auc: 0.59715 |  0:00:03s
epoch 8  | loss: 0.43697 | val_auc: 0.60283 |  0:00:03s
epoch 9  | loss: 0.43244 | val_auc: 0.61772 |  0:00:03s
epoch 10 | loss: 0.43183 | val_auc: 0.61259 |  0:00:04s
epoch 11 | loss: 0.42576 | val_auc: 0.62    |  0:00:04s
epoch 12 | loss: 0.42654 | val_auc: 0.62215 |  0:00:04s
epoch 13 | loss: 0.42733 | val_auc: 0.62683 |  0:00:05s
epoch 14 | loss: 0.41983 | val_auc: 0.635   |  0:00:05s
epoch 15 | loss: 0.41644 | val_auc: 0.65098 |  0:00:05s
epoch 16 | loss: 0.41917 | val_auc: 0.65055 |  0:00:06s
epoch 17 | loss: 0.41685 | val_auc: 0.65503 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 5/5 | Fold 1 AUC: 0.6760


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.68408 | val_auc: 0.48315 |  0:00:00s
epoch 1  | loss: 0.59066 | val_auc: 0.48434 |  0:00:00s
epoch 2  | loss: 0.51968 | val_auc: 0.50697 |  0:00:01s
epoch 3  | loss: 0.49065 | val_auc: 0.5198  |  0:00:01s
epoch 4  | loss: 0.47047 | val_auc: 0.52445 |  0:00:01s
epoch 5  | loss: 0.45802 | val_auc: 0.55821 |  0:00:02s
epoch 6  | loss: 0.44971 | val_auc: 0.56058 |  0:00:02s
epoch 7  | loss: 0.44301 | val_auc: 0.58295 |  0:00:02s
epoch 8  | loss: 0.42716 | val_auc: 0.59584 |  0:00:03s
epoch 9  | loss: 0.42675 | val_auc: 0.59336 |  0:00:03s
epoch 10 | loss: 0.42922 | val_auc: 0.60213 |  0:00:03s
epoch 11 | loss: 0.42053 | val_auc: 0.61757 |  0:00:04s
epoch 12 | loss: 0.41831 | val_auc: 0.61741 |  0:00:04s
epoch 13 | loss: 0.42054 | val_auc: 0.61882 |  0:00:04s
epoch 14 | loss: 0.41428 | val_auc: 0.62421 |  0:00:05s
epoch 15 | loss: 0.40964 | val_auc: 0.62237 |  0:00:05s
epoch 16 | loss: 0.41064 | val_auc: 0.63579 |  0:00:05s
epoch 17 | loss: 0.41095 | val_auc: 0.63399 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.69717 | val_auc: 0.49545 |  0:00:00s
epoch 1  | loss: 0.56443 | val_auc: 0.49059 |  0:00:00s
epoch 2  | loss: 0.50703 | val_auc: 0.50395 |  0:00:01s
epoch 3  | loss: 0.4828  | val_auc: 0.52756 |  0:00:01s
epoch 4  | loss: 0.4676  | val_auc: 0.53237 |  0:00:01s
epoch 5  | loss: 0.4549  | val_auc: 0.54366 |  0:00:02s
epoch 6  | loss: 0.44998 | val_auc: 0.56046 |  0:00:02s
epoch 7  | loss: 0.43873 | val_auc: 0.55294 |  0:00:02s
epoch 8  | loss: 0.43733 | val_auc: 0.56676 |  0:00:03s
epoch 9  | loss: 0.43337 | val_auc: 0.58212 |  0:00:03s
epoch 10 | loss: 0.42854 | val_auc: 0.58406 |  0:00:03s
epoch 11 | loss: 0.42606 | val_auc: 0.58899 |  0:00:04s
epoch 12 | loss: 0.42815 | val_auc: 0.60006 |  0:00:04s
epoch 13 | loss: 0.41994 | val_auc: 0.61018 |  0:00:04s
epoch 14 | loss: 0.41802 | val_auc: 0.61199 |  0:00:05s
epoch 15 | loss: 0.42045 | val_auc: 0.61125 |  0:00:05s
epoch 16 | loss: 0.41628 | val_auc: 0.61514 |  0:00:05s
epoch 17 | loss: 0.41698 | val_auc: 0.62173 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


Config 5/5 | Fold 3 AUC: 0.6673


/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.68704 | val_auc: 0.46124 |  0:00:00s
epoch 1  | loss: 0.57415 | val_auc: 0.49837 |  0:00:00s
epoch 2  | loss: 0.5246  | val_auc: 0.49759 |  0:00:01s
epoch 3  | loss: 0.4814  | val_auc: 0.49869 |  0:00:01s
epoch 4  | loss: 0.45771 | val_auc: 0.51968 |  0:00:01s
epoch 5  | loss: 0.45434 | val_auc: 0.54178 |  0:00:02s
epoch 6  | loss: 0.44498 | val_auc: 0.54647 |  0:00:02s
epoch 7  | loss: 0.44504 | val_auc: 0.55632 |  0:00:02s
epoch 8  | loss: 0.43002 | val_auc: 0.54732 |  0:00:03s
epoch 9  | loss: 0.42686 | val_auc: 0.55683 |  0:00:03s
epoch 10 | loss: 0.424   | val_auc: 0.57313 |  0:00:03s
epoch 11 | loss: 0.42164 | val_auc: 0.57888 |  0:00:04s
epoch 12 | loss: 0.42717 | val_auc: 0.58678 |  0:00:04s
epoch 13 | loss: 0.41916 | val_auc: 0.5885  |  0:00:04s
epoch 14 | loss: 0.41545 | val_auc: 0.5969  |  0:00:05s
epoch 15 | loss: 0.41254 | val_auc: 0.60095 |  0:00:05s
epoch 16 | loss: 0.40977 | val_auc: 0.61727 |  0:00:05s
epoch 17 | loss: 0.41403 | val_auc: 0.61521 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)
/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.66446 | val_auc: 0.47572 |  0:00:00s
epoch 1  | loss: 0.58453 | val_auc: 0.5261  |  0:00:00s
epoch 2  | loss: 0.52168 | val_auc: 0.51506 |  0:00:01s
epoch 3  | loss: 0.49088 | val_auc: 0.52245 |  0:00:01s
epoch 4  | loss: 0.47655 | val_auc: 0.55146 |  0:00:01s
epoch 5  | loss: 0.45511 | val_auc: 0.56717 |  0:00:02s
epoch 6  | loss: 0.45085 | val_auc: 0.56885 |  0:00:02s
epoch 7  | loss: 0.44227 | val_auc: 0.57791 |  0:00:02s
epoch 8  | loss: 0.43544 | val_auc: 0.57805 |  0:00:03s
epoch 9  | loss: 0.43173 | val_auc: 0.59248 |  0:00:03s
epoch 10 | loss: 0.43545 | val_auc: 0.60716 |  0:00:03s
epoch 11 | loss: 0.42593 | val_auc: 0.61117 |  0:00:04s
epoch 12 | loss: 0.42431 | val_auc: 0.62241 |  0:00:04s
epoch 13 | loss: 0.42687 | val_auc: 0.62145 |  0:00:04s
epoch 14 | loss: 0.41874 | val_auc: 0.61295 |  0:00:05s
epoch 15 | loss: 0.41389 | val_auc: 0.6404  |  0:00:05s
epoch 16 | loss: 0.41466 | val_auc: 0.6258  |  0:00:06s
epoch 17 | loss: 0.41713 | val_auc: 0.62995 |  0

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


## Summary of Results from CV

In [39]:
# sort best to worst
results_sorted = sorted(results, key=lambda x: x[2], reverse=True)

print("===== Summary (best first) =====")
for ci, cfg, mean_auc, std_auc in results_sorted:
    print(f"Config {ci}: AUC {mean_auc:.4f} ± {std_auc:.4f} | {cfg}")

===== Summary (best first) =====
Config 3: AUC 0.6702 ± 0.0094 | {'n_d': 32, 'n_a': 32, 'n_steps': 5, 'gamma': 1.8, 'lambda_sparse': 0.01, 'lr': 0.001}
Config 4: AUC 0.6692 ± 0.0080 | {'n_d': 32, 'n_a': 32, 'n_steps': 5, 'gamma': 1.5, 'lambda_sparse': 0, 'lr': 0.001}
Config 1: AUC 0.6653 ± 0.0048 | {'n_d': 32, 'n_a': 32, 'n_steps': 5, 'gamma': 1.5, 'lambda_sparse': 0.001, 'lr': 0.001}
Config 2: AUC 0.6653 ± 0.0131 | {'n_d': 32, 'n_a': 32, 'n_steps': 6, 'gamma': 1.5, 'lambda_sparse': 0.001, 'lr': 0.001}
Config 5: AUC 0.6642 ± 0.0116 | {'n_d': 64, 'n_a': 64, 'n_steps': 5, 'gamma': 1.5, 'lambda_sparse': 0.001, 'lr': 0.001}


## Pick Best Config from Cross Validation Results

In [42]:
best_ci, best_cfg, best_mean_auc, best_std = max(results, key=lambda x: x[2])
print(f"Best config = {best_ci} | CV AUC = {best_mean_auc:.4f} ± {best_std:.4f}")
print(best_cfg)

Best config = 3 | CV AUC = 0.6702 ± 0.0094
{'n_d': 32, 'n_a': 32, 'n_steps': 5, 'gamma': 1.8, 'lambda_sparse': 0.01, 'lr': 0.001}


## Train Final Model on All Training Data

In [44]:
# split the dataset into test and val again but doing it on a smaller val

# the validation set here simply determines when to stop training
# to reduce overfitting during final model training.
X_tr, X_val, y_tr, y_val = train_test_split(
    X, y, test_size=0.1, random_state=426, stratify=y
)

# build findal model using best hyperparameters selected from CV
final_model = TabNetClassifier(
    n_d=best_cfg["n_d"],
    n_a=best_cfg["n_a"],
    n_steps=best_cfg["n_steps"],
    gamma=best_cfg["gamma"],
    lambda_sparse=best_cfg["lambda_sparse"],
    optimizer_fn=torch.optim.AdamW,
    optimizer_params=dict(lr=best_cfg["lr"], weight_decay=1e-5),
    mask_type="entmax",
    device_name=device
)

# train final model with early stopping
final_model.fit(
    X_tr, y_tr,
    eval_set=[(X_val, y_val)],
    eval_metric=["auc"],
    max_epochs=100,
    patience=10,
    batch_size=1024,
    virtual_batch_size=128,
    drop_last=False
)

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/abstract_model.py:82: UserWarning: Device used : cuda
  warnings.warn(f"Device used : {self.device}")


epoch 0  | loss: 0.67866 | val_0_auc: 0.47299 |  0:00:00s
epoch 1  | loss: 0.57781 | val_0_auc: 0.47986 |  0:00:00s
epoch 2  | loss: 0.54047 | val_0_auc: 0.47548 |  0:00:01s
epoch 3  | loss: 0.504   | val_0_auc: 0.47056 |  0:00:01s
epoch 4  | loss: 0.48458 | val_0_auc: 0.49773 |  0:00:01s
epoch 5  | loss: 0.47658 | val_0_auc: 0.53323 |  0:00:02s
epoch 6  | loss: 0.46634 | val_0_auc: 0.54545 |  0:00:02s
epoch 7  | loss: 0.45431 | val_0_auc: 0.55454 |  0:00:03s
epoch 8  | loss: 0.45261 | val_0_auc: 0.5545  |  0:00:03s
epoch 9  | loss: 0.44459 | val_0_auc: 0.56165 |  0:00:03s
epoch 10 | loss: 0.4335  | val_0_auc: 0.58838 |  0:00:04s
epoch 11 | loss: 0.43596 | val_0_auc: 0.60983 |  0:00:04s
epoch 12 | loss: 0.43003 | val_0_auc: 0.62307 |  0:00:05s
epoch 13 | loss: 0.42734 | val_0_auc: 0.62005 |  0:00:05s
epoch 14 | loss: 0.42605 | val_0_auc: 0.62478 |  0:00:05s
epoch 15 | loss: 0.42501 | val_0_auc: 0.62593 |  0:00:06s
epoch 16 | loss: 0.42015 | val_0_auc: 0.62086 |  0:00:06s
epoch 17 | los

/usr/local/lib/python3.12/dist-packages/pytorch_tabnet/callbacks.py:172: UserWarning: Best weights from best epoch are automatically used!
  warnings.warn(wrn_msg)


## Test Set Results:

In [49]:
test_proba = final_model.predict_proba(X_test)[:, 1]
test_auc = roc_auc_score(y_test, test_proba)
print(f"TEST AUC: {test_auc:.4f}")

TEST AUC: 0.6489


## Confustion Matrix:

In [53]:
# Predict class labels
y_pred = final_model.predict(X_val)

# Compute confusion matrix
cm = confusion_matrix(y_val, y_pred)

print(cm)

[[1656    2]
 [ 275    2]]


In [60]:
y_proba = final_model.predict_proba(X_val)[:, 1]

precisions, recalls, thresholds = precision_recall_curve(y_val, y_proba)
f1_scores = 2 * (precisions * recalls) / (precisions + recalls + 1e-8)

best_idx = np.argmax(f1_scores)
best_threshold = thresholds[best_idx]

print("best threshold:", best_threshold)
print("best f1:", f1_scores[best_idx])

best threshold: 0.19179867
best f1: 0.35851851367918797


In [61]:
y_pred = (y_proba >= best_threshold).astype(int)
cm = confusion_matrix(y_val, y_pred)

print(cm)

[[1381  277]
 [ 156  121]]


In [63]:
pr_auc = average_precision_score(y_val, y_proba)
print("PR AUC:", pr_auc)

PR AUC: 0.28090036114472244


## Interpretation:

We can’t just look at the raw confusion matrix because the dataset is imbalanced (only 14.3% hits), which means a model can appear “good” simply by predicting mostly non-hits. At the default 0.5 threshold, the confusion matrix shows very few positive predictions (only 2 true positives), giving extremely low recall despite high overall accuracy — this is misleading because the model is essentially ignoring the minority class. After adjusting the threshold to maximize F1 (~0.19), the model captures about 44% of hits with ~30% precision, reflecting a much more meaningful trade-off. More importantly, the PR AUC of 0.281 is nearly double the baseline of 0.143 (the class prevalence), indicating that the model has genuine ranking power beyond random guessing. Overall, the model does learn real signal, but proper threshold tuning is essential to obtain practically useful performance in this imbalanced setting.